src/tasks/LargeFileUploadTask.ts

```diff
@@ -0,0 +1,353 @@
+    }
+    if (!parsableFactory) {
+      const error = new Error("Parsable factory is undefined, Please provide a valid parsable factory");
+      error.name = "Invalid Parsable Factory Error";
+      throw error;
+    }
+    if (uploadStream?.locked) {
+      throw new Error("Please provide stream value");
+    }
+    if (maxSliceSize <= 0) {
+      this.maxSliceSize = DefaultSliceSize;
+    }
+    this.parsableFactory = parsableFactory;
+
+    this.Session = this.extractSessionInfo(uploadSession);
+    this.rangesRemaining = this.getRangesRemaining(this.Session);
+    this.errorMappings = errorMappings;
+  }
+
+  /**
+   * Uploads the file in a sequential order by slicing the file in terms of ranges.
+   *
+   * @param {IProgress} [progress] - Optional progress receiver to report upload progress.
+   * @returns {Promise<UploadResult<T>>} - The result of the upload.
+   * @throws {Error} If the upload fails.
+   */
+  public async upload(progress?: IProgress): Promise<UploadResult<T>> {
+    const sliceRequests = this.getUploadSliceRequests();
+    for (const request of sliceRequests) {
+      try {
+        const uploadResult = await request.uploadSlice(this.uploadStream);
+        progress?.report(request.rangeEnd);
+        if (uploadResult?.itemResponse || uploadResult?.location) {
+          return uploadResult;
+        }
+      } catch (e) {
+        console.error(e);
+      }
+    }
+    throw new Error("Upload failed");
+  }
+
+  /**
+   * Uploads a slice with retry logic.
+   *
+   * @param {UploadSlice<T>} uploadSlice - The upload slice to be uploaded.
+   * @param {number} [maxTries=3] - The maximum number of retry attempts.
+   * @returns {Promise<UploadResult<T> | undefined>} - The result of the upload.
+   * @throws {Error} If the maximum number of retries is reached.
+   */
+  private async uploadWithRetry(uploadSlice: UploadSlice<T>, maxTries = 3): Promise<UploadResult<T> | undefined> {
+    let uploadTries = 0;
+    while (uploadTries < maxTries) {
+      try {
+        return await uploadSlice.uploadSlice(this.uploadStream);
+      } catch (e) {
+        console.error(e);
+      }
+      uploadTries++;
+
+      if (uploadTries < maxTries) {
+        // Exponential backoff
+        await this.sleep(2000 * (uploadTries + 1));
+      }
+    }
+
+    throw new Error("Max retries reached");
+  }
+
+  /**
+   * @public
+   * Resumes the current upload session
+   * @param progress
+   */
+  public async resume(progress?: IProgress): Promise<UploadResult<T>> {
+    await this.updateSession();
+    return this.upload(progress);
+  }
+
+  /**
+   * Refreshes the current upload session status by making a GET request to the upload URL.
+   * Updates the session expiration date, next expected ranges, and remaining ranges based on the response.
+   *
+   * @throws {Error} If the request fails.
+   */
+  public async updateSession() {
+    const requestInformation = new RequestInformation(HttpMethod.GET, this.Session.uploadUrl!);
```

Comment:
avoid bang operators, they lead to nasty bugs.
instead, validate the value, and throw an explicit error